ESQL: Support ST_DISJOINT

craigtaverner opened the pull request:
Following on from the ESQL spatial query work already done:

* https://github.com/elastic/elasticsearch/pull/104907
* https://github.com/elastic/elasticsearch/pull/106503

We add the fourth and final of the four Lucene-backed spatial search predicates: `ST_DISJOINT`. This is essentially the opposite of `ST_INTERSECTS` such that if two geometries intersect, they are not disjoint, and if they do not intersect, they are disjoint. The PostGIS documentation can be comared at https://postgis.net/docs/ST_Disjoint.html

craigtaverner committed:
WIP Started developing ST_DISJOINT

Initially based on ST_INTERSECTS

craigtaverner committed:
Update docs/changelog/107007.yaml

elasticsearchmachine commented:
Hi @craigtaverner, I've created a changelog YAML for you.

craigtaverner committed:
Fix functions list and add spatial point integration tests

craigtaverner committed:
More tests for shapes and cartesian-multigeoms

craigtaverner committed:
Some more tests to highlight issues with DISJOINT on cartesian point indices

craigtaverner committed:
Disable Lucene push-down for DISJOINT on cartesian point indices

craigtaverner committed:
Added docs for ST_DISJOINT

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

craigtaverner committed:
Support DISJOINT in the lucene-pushdown code for cartesian point indexes

craigtaverner committed:
Re-enable push-to-source for DISJOINT on cartesian_point indices

craigtaverner committed:
Fix docs example

craigtaverner committed:
Try fix internal docs links which are not being rendered

iverase commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/querydsl/query/SpatialRelatesQuery.java:
```diff
@@ -231,20 +243,29 @@ private static org.apache.lucene.search.Query pointShapeQuery(
-                // Should never be null, but can be an empty geometry
-                return new MatchNoDocsQuery();
+                // Should never be null, but could be an empty geometry
+                return new MatchNoDocsQuery("Empty geometry");
```
Is this correct for disjoint?

iverase reviewed (changes requested):
Just a small comment for empty geometries that I think needs addressing

craigtaverner committed:
Fixed disjoint on empty geometry

craigtaverner committed:
Added tests on empty linestring, and changed lucene push-down to exception

In lucene code only LineString can be empty, but in Elasticsearch even that is not allowed, resulting in parsing errors. So we cannot get to this code in the lucene push-down and now throw an error instead. The tests now assert on the warnings.

Note that for any predicate DISJOINT and INTERSECTS alike, the predicate fails, because the parsing error results in null, the function returns null, the predicate interprets this as false, and no documents match. This null-in-null-out rule means that DISJOINT and INTERSECTS give the same answer on invalid geometries.
